### ISSUE TITLE
Add release tags and GitHub releases

### ISSUE THREAD

**Issue by matthewfeickert** [mg5amcnlo/mg5amcnlo#10]
Moving from the LaunchPad question [Status of "transition to git" Blueprint given madgraph5 GitHub Organization](https://answers.launchpad.net/mg5amcnlo/+question/693060) to here, can [release tags](https://github.com/mg5amcnlo/mg5amcnlo/tags) from LaunchPad be ported over to the relevant commits and can associated [GitHub releases](https://github.com/mg5amcnlo/mg5amcnlo/releases) be made for them?

I appreciate that this would be a lot of manual work, but even just adding the `MG5_aMC` `v3.4.0` release tag on the relevant commit and making a GitHub release and then slowly over time adding older releases would be a big help.

**Comment by oliviermattelaer** [comment]
I guess that we are still learning github.

It is true that for the release, we are still using the launchpad release mainly because all our CI/CD are based on it. 
Actually the tagging of "old" commit within git will not work, since the conversion script lead to invalid state that we did fix as one additional commit fixing the stuff.

Now we could have done it for 2.9.11 which was released from git.

**Comment by oliviermattelaer** [comment]
> Also, given the commit message of [f764f8a](https://github.com/mg5amcnlo/mg5amcnlo/commit/f764f8aa33d0f37308def224e555b13952f527cc), would it be correct to assume that this is the commit that corresponds to `v3.4.0`?

I would say no. That commit does not correspond to 3.4.0 but correspond to the auto-generated history in git as imported from bazaar. It means that such commit is likely "not working". Everything older (including) 3.4.0 or 2.9.10 need to be using launchpad/bazaar.

So only 2.9.11 can be associate to a tag in github for the moment.

> Will it be possible to start making releases from the Git repo in the future?

Doing tag makes a lot of sense, for releases I do not know.
Not fundamentally opposed but this just made double counting with the launchpad website so likely low interest.

**Comment by matthewfeickert** [comment]
Thanks for clarifying RE: f764f8aa33d0f37308def224e555b13952f527cc.

> Doing tag makes a lot of sense, for releases I do not know. Not fundamentally opposed but this just made double counting with the launchpad website so likely low interest.

As a user it is a bit confusing to have development be on one platform but official releases be on another. (To be clear, I think I understand because you've explained it to me, but I wouldn't assume other users who haven't read this Issue thread would.). So until everything is migrated over, simply marking a release in GitHub seems like it would be a helpful action.

> we are still using the launchpad release mainly because all our CI/CD are based on it.

So is the only thing that is keeping full migration to GitHub getting CI up and running with GitHub Actions? Or are there additional things that need to get migrated but are difficult?

Is the dev team open to moving the CI to GitHub Actions? If so, and if the main blocker is setting up the CI, are there developer docs that explain how to setup the dependencies/environment that are accepted as valid for the test suite and how to run the tests other than this section of the README?

https://github.com/mg5amcnlo/mg5amcnlo/blob/b198aa8eaa346bef9fd0fce7e3da302d6a847b53/README.developer#L33-L43

**Comment by oliviermattelaer** [comment]
In principle starting from 3.5.0 all official release would have two tags:
1. the version number of the bzr equivalent version number (since we still need those for a lot of feature and this will allow us at some point to emulate bzr within github) (those will start with r then 3 or 4 digit)
2. The actual human readable version (those will start with a v)

For LTS version, I guess the same will be put in place (but not yet ready)
same format will be applied but they will be prefix (or postfis) with "LTS"

For "release", I do not plan to put them on this platform for the moment.

**Comment by matthewfeickert** [comment]
:wave: @oliviermattelaer following our conversation at CHEP 2024 about the idea of me attempting to package `mg5amcnlo` on conda-forge, I wanted to revisit the idea of tags/releases on GitHub:

Packages on conda-forge need to define a [`source` map](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#source-section) that contains a `url` of a source archive and a `sha256` checksum ([conda-forge enforces the `sha256` choice](https://github.com/conda-forge/staged-recipes/blob/3015aa8ed615b39d7ac18445149c0bb6e310d01b/recipes/example/meta.yaml#L23) here) of that archive for build reproducibility reasons so that the CI infrastructure conda-forge employs can do automatic rebuilds.

As the Launchpad downloads site (https://launchpad.net/mg5amcnlo/+download) is not used as an archive of releases (c.f. https://github.com/mg5amcnlo/mg5amcnlo/issues/16#issuecomment-1185851411) and patch releases will be overwritten (e.g. On 2024-10-28 I can download [`v3.5.x`](https://launchpad.net/mg5amcnlo/3.0/3.5.x) patch release `v3.5.6` from https://launchpad.net/mg5amcnlo/3.0/3.5.x/+download/MG5_aMC_v3.5.6.tar.gz but I can not download `v3.5.5` as that has been removed) the Launchpad site can not be used for the `source` `url` field.

It is generally preferable to use a version controlled source repository on a public hosting service, like GitHub, so given

> In principle starting from 3.5.0 all official release would have two tags:
> 
> 1. the version number of the bzr equivalent version number (since we still need those for a lot of feature and this will allow us at some point to emulate bzr within github) (those will start with r then 3 or 4 digit)
> 2. The actual human readable version (those will start with a v)

is it now correct that the repository at the [`v3.5.6` tag](https://github.com/mg5amcnlo/mg5amcnlo/releases/tag/v3.5.6) should be equivalent in content to the `v3.5.6` release on Launchpad?

**edit:** Answering my own question: No, given

```console
$ curl -sLO https://github.com/mg5amcnlo/mg5amcnlo/archive/refs/tags/v3.5.6.tar.gz
$ curl -sLO https://launchpad.net/mg5amcnlo/3.0/3.5.x/+download/MG5_aMC_v3.5.6.tar.gz
$ tar -xzf v3.5.6.tar.gz
$ tar -xzf MG5_aMC_v3.5.6.tar.gz
$ diff --brief --recursive mg5amcnlo-3.5.6/ MG5_aMC_v3_5_6/
Only in MG5_aMC_v3_5_6/bin: .compile.py
Only in mg5amcnlo-3.5.6/bin: compile.py
Only in mg5amcnlo-3.5.6/bin: create_aloha_release.py
Only in mg5amcnlo-3.5.6/bin: create_release.py
Only in MG5_aMC_v3_5_6/input: .autoupdate
Only in MG5_aMC_v3_5_6/input: default_run_card_lo.dat
Only in MG5_aMC_v3_5_6/input: default_run_card_nlo.dat
Only in MG5_aMC_v3_5_6/input: mg5_configuration.txt
Only in MG5_aMC_v3_5_6/: proc_card.dat
Only in MG5_aMC_v3_5_6/: README
Only in mg5amcnlo-3.5.6/: README.developer
Only in mg5amcnlo-3.5.6/: README.release
diff: mg5amcnlo-3.5.6/Template/NLO/MCatNLO/include/pineappl_maxproc.inc: No such file or directory
diff: MG5_aMC_v3_5_6/Template/NLO/MCatNLO/include/pineappl_maxproc.inc: No such file or directory
Only in MG5_aMC_v3_5_6/vendor: collier.tar.gz
Only in MG5_aMC_v3_5_6/vendor: ninja.tar.gz
Only in MG5_aMC_v3_5_6/vendor: OfflineHEPToolsInstaller.tar.gz
Only in MG5_aMC_v3_5_6/vendor: SMWidth.tar.gz
diff: mg5amcnlo-3.5.6/vendor/SudGen/NNPDFDriver.f: No such file or directory
diff: MG5_aMC_v3_5_6/vendor/SudGen/NNPDFDriver.f: No such file or directory
$
```

(The `No such file or directory` bits are raised due to symlinks and so can be ignored.)

Are these differences meaningful? Or would a distribution created from the GitHub tags be viewed by you as canonical and complete?

**Comment by oliviermattelaer** [comment]
Hi Matthew,

The link for LTS releases are stable:
https://launchpad.net/mg5amcnlo/lts/2.9.x/+download/MG5_aMC_v2.9.21.tar.gz
https://launchpad.net/mg5amcnlo/lts/lts.3.5.x/+download/MG5_aMC_v3.5.6.tar.gz
Version which are not LTS does not have such type of link (including 3.5.5 and 3.6.0)

But the link https://launchpad.net/mg5amcnlo/3.0/3.5.x/+download/MG5_aMC_v3.5.6.tar.gz 
will indeed not stay as soon as 3.5.7 will be released.

```
is it now correct that the repository at the [v3.5.6 tag](https://github.com/mg5amcnlo/mg5amcnlo/releases/tag/v3.5.6) should be equivalent in content to the v3.5.6 release on Launchpad?
```

The tarball are not the same and this will impact performance since the code will automatically setting itself in debug mode 
if you use any of the tarball created by github.

Also the lack of some of the configuration file in such release can sometimes create problem depending on the way you run the code.

Finally, some information message (like detection that a bug fix has been released for the current version) will not work since the associated meta-data will be lacking.

```
Are these differences meaningful? Or would a distribution created from the GitHub tags be viewed by you as canonical and complete?
```

Yes this is the difference that I do expect since those file are added by our release script.

**Comment by matthewfeickert** [comment]
Thanks @oliviermattelaer.

As I assume that your view is _not_ that people should only ever use LTS releases (otherwise I don't think you would make non-LTS releases public) would you be open to either of the following?

* **Idea 1** (more work initially, better long run workflow with automation): It seems that the release processes

https://github.com/mg5amcnlo/mg5amcnlo/blob/55a291dc8def762938a964cc1343c6d76af5b843/bin/create_release.py#L18-L30

https://github.com/mg5amcnlo/mg5amcnlo/blob/55a291dc8def762938a964cc1343c6d76af5b843/bin/create_aloha_release.py#L18-L32

are more like build steps given that archives that aren't Git submodules are getting downloaded and vendored

https://github.com/mg5amcnlo/mg5amcnlo/blob/55a291dc8def762938a964cc1343c6d76af5b843/bin/create_release.py#L330

and files are getting copied around and renamed

https://github.com/mg5amcnlo/mg5amcnlo/blob/55a291dc8def762938a964cc1343c6d76af5b843/bin/create_release.py#L283-L292

. But as you can use GitHub Actions to create workflows that will automatically create and upload release artifacts, all of this could be automated on GitHub to generate release tarballs automatically upon creation of a GitHub release.

I'm not sure how much you've migrated from Bazaar/Launchpad to Git/GitHub since 2022 (https://github.com/mg5amcnlo/mg5amcnlo/issues/10#issuecomment-1167776274), but do you already have a CI workflow that does the full releases on Launchpad with Bazaar?

* **Idea 2** (less work, but much more a patch than a solution): Would you be willing to manually upload the tarball of each MadGraph release from Launchpad to the corresponding tag so that GitHub can also serve as an archive of the releases?
   - e.g. Attach https://launchpad.net/mg5amcnlo/3.0/3.5.x/+download/MG5_aMC_v3.5.6.tar.gz to https://github.com/mg5amcnlo/mg5amcnlo/releases/tag/v3.5.6 as additional post-tag assets by [editing the release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository#editing-a-release).

**Comment by jpivarski** [comment]
> But the link https://launchpad.net/mg5amcnlo/3.0/3.5.x/+download/MG5_aMC_v3.5.6.tar.gz
> will indeed not stay as soon as 3.5.7 will be released.

Wait, the old release just disappears?

I don't think it's necessary to have the releases in GitHub, and it doesn't bother me too much to have development in one place and hosting in another, but are you saying that if someone does some physics work with version 1.2.3 and then wants to repeat it years later, they won't be able to download version 1.2.3 anymore? Or can you download it, but not from the same place?

**Comment by matthewfeickert** [comment]
> but are you saying that if someone does some physics work with version 1.2.3 and then wants to repeat it years later, they won't be able to download version 1.2.3 anymore? Or can you download it, but not from the same place?

@jpivarski AFAIK, they can't download it. If `1.2.3` isn't a LTS release it is just overwritten and gone given the discussion from 2022 on https://github.com/mg5amcnlo/mg5amcnlo/issues/16#issuecomment-1185261810. :(

**Comment by oliviermattelaer** [comment]
Only version which have a bug fixing release lose their link.
So 1.2.3 will only be lost if a bug fixing 1.2.4 is released.
If we do a feature release (1.3.0) then 1.2.3 will be kept.

For having the tarball to github, I would have to think, but having multiple method (CI/script) to create the tarball is a no-go.
We should have a single source of truth. Now the github action can download the link from launchpad.

**Comment by jpivarski** [comment]
Not being facetious: a physics analysis might be depending on a bug in 1.2.3. Specifically, someone might be trying to answer the question, "does the weird artifact in my plots come from the bug that was fixed between 1.2.3 and 1.2.4?" and having access to 1.2.3 is the only way to fully answer that question.

Maybe the archive of old versions can be on another site, like deadsnakes for Python, but not being able to access it anywhere is dangerous. (It may be annoying to change download URLs or it may be hard to discover that this is what needs to be done, but making it possible is the first big step.)

Is it currently possible to access the source code with git (or similar) and checkout a tag for 1.2.3? (Or look up a commit hash for 1.2.3 and use that?) That's almost as good as having a tarball.

**Comment by matthewfeickert** [comment]
> We should have a single source of truth. Now the github action can download the link from launchpad.

@oliviermattelaer I agree that having a single source of truth is a good idea. :+1: Can you remind me though why Launchpad is still that single source of truth, given that on 2022-05-24 in [Status of "transition to git" Blueprint given madgraph5 GitHub Organization](https://answers.launchpad.net/mg5amcnlo/+question/693060) you mentioned that

> launchpad repo are still used for the moment for all branch that have started their development before 3.4.0.
As soon as they branch will be merged with 3.4.0, I will move them on git/github and they will continue to grow on github.

and we're now on [`v3.6.0`](https://github.com/mg5amcnlo/mg5amcnlo/releases/tag/v3.6.0). From the [Launchpad release history timeline](https://launchpad.net/mg5amcnlo/+series) it looks like the LTS series has moved beyond `v2.9.x` to now be `v3.5.x`.

[![image](https://github.com/user-attachments/assets/fcd845a7-ae77-42a0-b48d-fa912fd6092a)](https://launchpad.net/mg5amcnlo/+series)

Though [LTS is DEAD long live to the new LTS](https://launchpad.net/mg5amcnlo/+announcement/54113) reads

> Long Term Stable 2.9.x has now an end of live: december 2025.
A new Long Term Stable release (3.5.x) is starting now to replace it.
Version 3.6.0 will be the new feature oriented release.

So does that mean that you won't fully move to GitHub until 2026?

To be clear, I ask this only to understand what the workflow is, not to try to challenge your choices and decisions.

**Comment by matthewfeickert** [comment]
> Is it currently possible to access the source code with git (or similar) and checkout a tag for 1.2.3? (Or look up a commit hash for 1.2.3 and use that?) That's almost as good as having a tarball.

Not with Git, as going back one release to [`v3.5.5`](https://github.com/mg5amcnlo/mg5amcnlo/releases/tag/v3.5.5) the `bin/create_release.py` script has an error in it:

```console
$ docker run --rm -ti python:3.12 /bin/bash
root@79819a3e72ce:/# python -m venv venv && . venv/bin/activate
(venv) root@79819a3e72ce:/# git clone --recursive https://github.com/mg5amcnlo/mg5amcnlo.git && cd mg5amcnlo
Cloning into 'mg5amcnlo'...
remote: Enumerating objects: 140733, done.
remote: Counting objects: 100% (355/355), done.
remote: Compressing objects: 100% (197/197), done.
remote: Total 140733 (delta 184), reused 295 (delta 150), pack-reused 140378 (from 1)
Receiving objects: 100% (140733/140733), 766.51 MiB | 11.72 MiB/s, done.
Resolving deltas: 100% (112363/112363), done.
root@79819a3e72ce:/mg5amcnlo# git checkout v3.5.5
Note: switching to 'v3.5.5'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 2acd31be9 one more pass trough ruff (ruff --select W605 --fix)
(venv) root@79819a3e72ce:/mg5amcnlo# python -m pip install --quiet six  # six is still required for some reason
(venv) root@79819a3e72ce:/mg5amcnlo# ./bin/create_release.py 
version = 3.5.5

WARNING: The release time information is : date = 2024-06-17

Do you want to continue anyway? (y/n)y
Traceback (most recent call last):
  File "/mg5amcnlo/./bin/create_release.py", line 124, in <module>
    print("cannot create tarball with auto-update outside of the main branch, detected branch (%s)" % branch)
                                                                                                      ^^^^^^
NameError: name 'branch' is not defined. Did you mean: 'MG_branch'?
(venv) root@79819a3e72ce:/mg5amcnlo# 
```

(This was fixed in PR https://github.com/mg5amcnlo/mg5amcnlo/pull/91 but that only went into [`v3.6.0`](https://github.com/mg5amcnlo/mg5amcnlo/blob/55a291dc8def762938a964cc1343c6d76af5b843/bin/create_release.py#L123-L124), so [`v3.5.6` has the same bug](https://github.com/mg5amcnlo/mg5amcnlo/blob/9ca5770ebe6732b631be89c8b6d778a3e8b86660/bin/create_release.py#L123-L124))

In https://github.com/mg5amcnlo/mg5amcnlo/issues/16#issuecomment-1185640718 it was mentioned that 

> one is force to use bzr.

but that was in 2022, so I'm not sure if you should be able to using Git now (if you manually patch the bugs).

Though given that the release process relies on vendoring unversioned tarballs

https://github.com/mg5amcnlo/mg5amcnlo/blob/55a291dc8def762938a964cc1343c6d76af5b843/bin/create_release.py#L346-L353

selecting a release tag won't get you a reproducible build.

Also 

https://github.com/mg5amcnlo/mg5amcnlo/blob/55a291dc8def762938a964cc1343c6d76af5b843/bin/create_release.py#L348

404s

```console
$ curl --head --silent https://bitbucket.org/peraro/ninja/downloads/ninja-latest.tar.gz | grep HTTP
HTTP/2 404
```

**Comment by oliviermattelaer** [comment]
I would say that those facts are completely independent of our usage of launchpad/github, they only impacts on which version we provide support and provide bug fixing update released.

For the rest of the transition:
- we do not use bazaar anymore (even for some CI so here the transition to git is finished)
- we still use launchpad quite a lot for user support: tarball, question and bug report and do not plan to drop our usage of launchpad anytime soon (even if user support can be done via github as well)


So does that mean that you won't fully move to GitHub until 2026?

We do not have any plan for additional transition from launchpad to github for the moment.
The next transition will be our wiki page/documentation (but it is not clear yet if this will go to a gitlab/github or something else)




On 28 Oct 2024, at 22:50, Matthew Feickert ***@***.***> wrote:



We should have a single source of truth. Now the github action can download the link from launchpad.

@oliviermattelaer<https://github.com/oliviermattelaer> I agree that having a single source of truth is a good idea. 👍 Can you remind me though why Launchpad is still that single source of truth, given that on 2022-05-24 in Status of "transition to git" Blueprint given madgraph5 GitHub Organization<https://answers.launchpad.net/mg5amcnlo/+question/693060> you mentioned that

launchpad repo are still used for the moment for all branch that have started their development before 3.4.0.
As soon as they branch will be merged with 3.4.0, I will move them on git/github and they will continue to grow on github.

and we're now on v3.6.0<https://github.com/mg5amcnlo/mg5amcnlo/releases/tag/v3.6.0>. From the Launchpad release history timeline<https://launchpad.net/mg5amcnlo/+series> it looks like the LTS series has moved beyond v2.9.x to now be v3.5.x.

<https://launchpad.net/mg5amcnlo/+series>image.png (view on web)<https://github.com/user-attachments/assets/fcd845a7-ae77-42a0-b48d-fa912fd6092a>

Though LTS is DEAD long live to the new LTS<https://launchpad.net/mg5amcnlo/+announcement/54113> reads

Long Term Stable 2.9.x has now an end of live: december 2025.
A new Long Term Stable release (3.5.x) is starting now to replace it.
Version 3.6.0 will be the new feature oriented release.

So does that mean that you won't fully move to GitHub until 2026?

To be clear, I ask this only to understand what the workflow is, not to try to challenge your choices and decisions.
